zebra-consensus/src/primitives/halo2/tests.rs

```diff
@@ -0,0 +1,146 @@
+//! Tests for verifying simple Halo2 proofs with the async verifier
+
+use std::convert::TryInto;
+
+use futures::stream::{FuturesUnordered, StreamExt};
+use tower::ServiceExt;
+
+use halo2::{arithmetic::FieldExt, pasta::pallas};
+use orchard::{
+    builder::Builder,
+    bundle::Flags,
+    circuit::ProvingKey,
+    keys::{FullViewingKey, SpendingKey},
+    value::NoteValue,
+    Anchor, Bundle,
+};
+use rand::rngs::OsRng;
+
+use zebra_chain::orchard::ShieldedData;
+
+use crate::primitives::halo2::*;
+
+lazy_static::lazy_static! {
+    pub static ref PROVING_KEY: ProvingKey = ProvingKey::build();
+}
+
+async fn verify_orchard_halo2_proofs<V>(
+    verifier: &mut V,
+    shielded_data: Vec<ShieldedData>,
+) -> Result<(), V::Error>
+where
+    V: tower::Service<Item, Response = ()>,
+    <V as tower::Service<Item>>::Error: std::convert::From<
+        std::boxed::Box<dyn std::error::Error + std::marker::Send + std::marker::Sync>,
+    >,
```

Comment:
Optional suggestion: I think this could be shortened by leveraging Rust's prelude:

```rust
    <V as tower::Service<Item>>::Error: From<Box<dyn std::error::Error + Send + Sync>>,
```
It can be shortened further using `BoxError` (either from `crate` or `tower`):

```suggestion
    <V as tower::Service<Item>>::Error: From<tower::BoxError>,
```

Reply:
Thanks! Done in 524b964f